보드 전체 조회 API 구현

Park-SooYeon opened the pull request:
# 보드 전체 조회 API 구현

## 📎 해당 이슈

이슈 링크 (#17 )

## 🛠 구현 내용 

- `GET /api/board`의 요청을 처리하는 API 구현
- 요청 처리를 위한 Service, Router 추가

## 🙋‍ 리뷰어 참고 사항 

- 테스트 할때마다 오류가 나서 아직 테스트 코드를 구현하지 못하였습니다.. 좀 더 공부하고 추가하겠습니다..테스트 코드 작성 많이 어렵네요ㅠ

Park-SooYeon committed:
[FEAT] Board 조회에 사용할 service 함수 추가

- 내가 가진 보드만 조회하는 getMyBoards 함수 추가
- 초대 받은 보드만 조회하는 getInvitedBoards 함수 추가

Park-SooYeon committed:
[FEAT] board api 사용을 위한 라우터 설정 추가

- board api의 요청 경로를 설정하는 board router 추가

Park-SooYeon committed:
Merge branch 'develop/web' into feature/get-boards-api

- develop/web 업데이트된 사항 pull

Park-SooYeon committed:
[FEAT] board router에 jwt 권한 확인 추가

- board router에 jwt 권한 확인 추가

Park-SooYeon committed:
[FIX] 응답 결과가 api 명세와 불일치하는 문제 수정

- jwt에 저장된 user id를 사용하여 board 조회하도록 변경
- 응답 결과가 api 명세와 불일치하는 문제 수정

youngxpepp commented on server/src/router/BoardRouter.js:
```diff
@@ -0,0 +1,15 @@
+        const boardService = BoardService.getInstance();
+        const { id } = req.user;
+        const data = await boardService.getBoardsByUserId(id);
+        res.status(200).json({ ...data });
```
```suggestion
        res.status(200).json(data);
```
data를 풀었다가 다시 객체로 묶어주는 것보단 data를 바로 전달하는게 좋아보여요.
의도하신건가욥?

Park-SooYeon replied:
data를 풀어주지 않으면 응답 json이 아래와 같이 한 뎁스가 더 나오더라구요..
그리고 api 명세서에 적은 결과와도 달라서 어쩔수없이 풀어주었습니다.
```
data : {
    myBoards: [],
    invitedBoards: []
}
```

차라리 ...을 사용하지 않고 좀 더 명시적으로 아래와 같은 코드를 작성하는게 더 좋을 수도 있을 것 같다는 생각이 드네요.🤔
```javascript
    const { myBoards, invitedBoards } = await boardService.getBoardsByUserId(id);
    res.status(200).json({ myBoards, invitedBoards });
```

youngxpepp replied:
``` javascript
res.json(data); 
```
이렇게 하신거 맞을까요? 아래와 같이 테스트할 땐 뎁스가 안 나옵니당😅
![image](https://user-images.githubusercontent.com/47842964/100092749-eca45a00-2e99-11eb-980a-b6b9853ebe3c.png)

Park-SooYeon replied:
아 생각해보니 `res.json({data});` 이렇게 썼었었네요.. 그러니 뎁스가 하나 더 생길 수 밖에 없었네요..ㅠ
수정하겠습니다; ㅁ;

Park-SooYeon replied:
commit eb06294
리뷰해주신 부분 모두 수정했습니다:)

youngxpepp commented on server/src/service/BoardService.js:
```diff
@@ -10,4 +11,32 @@ export class BoardService extends BaseService {
+            .createQueryBuilder('board')
+            .select('board.id')
+            .addSelect('board.title')
+            .innerJoin('board.invitations', 'invitation', 'invitation.user=:userId', { userId: id })
```
```suggestion
            .innerJoin('board.invitations', 'a', 'a.user_id=:userId', { userId: id })
```
테스트를 돌려보지 않아서 확실치 않지만, 위처럼 해야지 동작할 것 같아요.
Join 성능을 생각했을 때, board 보다는 invitation을 먼저 조회하는 것이 더 빠를 것 같다는 생각이 듭니다.
board를 먼저 조회하게 되면 invitation이 없는 board도 불필요하게 검색할 것 같아요.

Park-SooYeon replied:
model 폴더에서 정의한 엔티티의 속성대로 컬럼명이 따라가는 것으로 이해를 했었습니다.🤔
invitation 엔티티에서 user_id 컬럼이 user로 정의되어 있어 user로 사용했는데 이거는 쿼리문을 한번 확인해봐야겠네요.

Park-SooYeon replied:
일단 제가 실행해보았을 때, 제가 작성한 것과 건홍님이 작성한 것 모두 아래와 같은 동일한 쿼리가 만들어지더라구요.
```sql
SELECT "board"."id" AS "board_id", "board"."title" AS "board_title" FROM "board" "board"
INNER JOIN "invitation" "invitation" ON "invitation"."board_id"="board"."id" AND ("invitation"."user_id"=?)
```
신기하네요. typeORM👀
쿼리는 동일하지만 코드상으로 건홍님 코드가 컬럼명이 좀 더 직관적인 것 같다는 느낌이 들어서 join 순서 변경하면서 수정하도록 하겠습니다.

Park-SooYeon replied:
join 성능에 대한 말씀을 하셔서 inner join 성능에 대해서 이래저래 구글링 해보았습니다.
아래 글에 의하면 MySQL의 경우, 두개의 테이블 inner join은 테이블 순서에 상관없이 성능이 동일할 것이라는 생각이 드네요.
[MySQL 테이블 조인](https://12bme.tistory.com/165)

다만 세개 이상의 테이블을 inner join하는 경우에는 join 순서가 성능에 영향을 미치는 것 같네요.
[쿼리 최적화](https://medium.com/watcha/%EC%BF%BC%EB%A6%AC-%EC%B5%9C%EC%A0%81%ED%99%94-%EC%B2%AB%EA%B1%B8%EC%9D%8C-%EB%B3%B4%EB%8B%A4-%EB%B9%A0%EB%A5%B8-%EC%BF%BC%EB%A6%AC%EB%A5%BC-%EC%9C%84%ED%95%9C-7%EA%B0%80%EC%A7%80-%EC%B2%B4%ED%81%AC-%EB%A6%AC%EC%8A%A4%ED%8A%B8-bafec9d2c073) -> 6번 문항이 inner join 순서에 대한 얘기네요!

건홍님 의견은 어떠신가요?
DB 성능 고려하기..헷갈리네요..😵
MySQL은 join 방식이 NL 방식으로 고정이라고 들었는데..
오라클의 경우에는 여러 join 방식을 제공한다고 해서 그렇다면 오라클에서는 join 순서에 따라 더 많은 성능 차이가 나려나요.. 궁금해지네요🤔

youngxpepp replied:
좋은 글 감사합니다. 도움이 많이 됐어요👍
`("invitation"."user_id"=?)` 와 같은 추가적인 작업이 있어서 많이 헷갈렸는데 시원하게 해결됐습니다.
나중에 board_id와 user_id를 다중 인덱스 걸면 더 빠르게 찾을 수 있을 듯 합니다!

youngxpepp commented on server/src/service/BoardService.js:
```diff
@@ -10,4 +11,32 @@ export class BoardService extends BaseService {
+    async getBoardsByUserId(userId) {
+        const data = {};
+
+        data.myBoards = await this.getMyBoards(userId);
+        data.invitedBoards = await this.getInvitedBoards(userId);
```
```suggestion
        const promises = [this.getMyBoards(userId), this.getInvitedBoards(userId)];
        const [myBoards, invitedBoards] = await Promise.all(promises);
        return { myBoards, invitedBoards };
```
nodejs의 비동기 IO를 최대한 활용해보는건 어떨까요? 
37번 라인과 38번 라인은 동기적으로 (순차적으로) 실행되지 않아도 되는 로직이라서 동시에 요청을 보내면 효율적일 것 같다고 생각합니다.

Park-SooYeon replied:
아직은 비동기가 친숙하지 않아서 그런 생각은 하지 못했네요.
좋은 의견 감사합니다!:+1:

youngxpepp reviewed (changes requested):
TypeORM 처음 쓰시는 것 맞나요? 잘하셨습니당
다만, 테스트 로직이 없는게 좀 아쉬워요ㅠㅠ 응답은 확인 해보셨을까요?

dh00n reviewed (approved):
👏

Park-SooYeon committed:
[FIX] 리뷰 피드백 반영

- json 응답 코드 변경
- 초대 받은 보드를 불러올 때 사용하는 inner join문 컬럼명 수정
- 보드 전체 조회 로직을 비동기적으로 변경

Park-SooYeon committed:
[TEST] GET /api/board API 테스트 코드 추가

- GET /api/board API 테스트 코드 추가

youngxpepp reviewed (approved):
고생하셨습니다!
